File: solutions/advanced/DP on a Tree/MaximalPointPath.cpp
/*
    Maximal Point Path

*/ 
#include <iostream>
#include <vector>
using namespace std;
const int N = 500000+21;
vector<pair<int, int> > graph[N];
int dp[N];
int vis[N];
int points[N];
int n, a, b, c;
int ans;
void dfs(int v)
{
    vis[v] = 1;
    int ans1 = 0, ans2 = 0;
    for (int i=0; i<graph[v].size(); i++)
    {
        int curr = graph[v][i].first;
        int weight = graph[v][i].second;

        if (!vis[curr])
        {
            dfs(curr);
        }
        int cost = dp[curr] - weight;

        if (cost >= ans1)
        {
            ans2 = ans1;
            ans1 = cost;
        }
        else if (cost >= ans2)
            ans2 = cost;
    }
    ans = max(ans, ans1 + ans2 + points[v]);

    dp[v] = ans1 + points[v];
}
int main()
{
    cin>>n;
    for (int i=1; i<=n; i++)
        cin>>points[i];
    
    for (int i=1; i<n; i++)
    {
        cin>>a>>b>>c;
        graph[a].push_back({b, c});
        graph[b].push_back({a, c});
    }

    dfs(1);
    
    cout<<ans<<endl;
}

File: solutions/advanced/DP on a Tree/MaximalPointPath.py
n = int(input())
graph = [[] for _ in range(n + 1)]
dp = [0] * (n + 1)
vis = [0] * (n + 1)
points = [0] * (n + 1)
ans = 0

# Read point values
for i in range(1, n + 1):
    points[i] = int(input())

# Read edges (a, b, cost)
for _ in range(n - 1):
    a, b, c = map(int, input().split())
    graph[a].append((b, c))
    graph[b].append((a, c))

def dfs(v):
    global ans
    vis[v] = 1
    max1 = 0
    max2 = 0

    for neighbor, weight in graph[v]:
        if not vis[neighbor]:
            dfs(neighbor)
        cost = dp[neighbor] - weight
        if cost >= max1:
            max2 = max1
            max1 = cost
        elif cost > max2:
            max2 = cost

    ans = max(ans, max1 + max2 + points[v])
    dp[v] = max1 + points[v]

dfs(1)
print(ans)
